crates/symbolicator/src/services/symbolication.rs

```diff
@@ -1113,22 +1113,104 @@ fn symbolicate_frame(
     Ok(rv)
 }
 
+/// Stacktrace related Metrics
+///
+/// This gives some metrics about the quality of the stack traces included
+/// in a symbolication request. See the individual members for more information.
+#[derive(Default)]
+struct StacktraceMetrics {
+    /// A truncated stack trace is one that does not end in a
+    /// well known thread base.
+    truncated_traces: u64,
+
+    /// This indicated a stack trace that has at least one bad frame
+    /// from the below categories.
+    bad_traces: u64,
+
+    /// Frames that were scanned.
+    ///
+    /// These are frequently wrong and lead to bad and incomplete stack traces.
+    /// We can improve (lower) these numbers by having more usable CFI info.
+    scanned_frames: u64,
+
+    /// Unsymbolicated Frames.
+    ///
+    /// These may be the result of unavailable or broken debug info.
+    /// We can improve (lower) these numbers by having more usable debug info.
+    unsymbolicated_frames: u64,
+
+    /// Frames referencing unmapped memory regions.
+    ///
+    /// These may be the result of issues in the client-side module finder, or
+    /// broken debug-id information.
+    ///
+    /// We can improve this by fixing client-side implementations and having
+    /// proper debug-ids.
+    unmapped_frames: u64,
+}
+
+impl StacktraceMetrics {
+    fn update(&mut self, other: StacktraceMetrics) {
+        self.truncated_traces += other.truncated_traces;
+        self.bad_traces += other.bad_traces;
+        self.scanned_frames += other.scanned_frames;
+        self.unsymbolicated_frames += other.unsymbolicated_frames;
+        self.unmapped_frames += other.unmapped_frames;
+    }
+}
+
+/// Determine if the [`SymbolicatedFrame`] is likely to be a thread base.
+///
+/// This is just a heuristic that matches the function to well known thread entry points.
+fn is_likely_base_frame(frame: &SymbolicatedFrame) -> bool {
+    let function = match frame
+        .raw
+        .function
+        .as_deref()
+        .or_else(|| frame.raw.symbol.as_deref())
+    {
+        Some(f) => f,
+        None => return false,
+    };
+
+    // C start/main
+    if matches!(function, "main" | "start") {
+        return true;
+    }
+
+    // Windows and posix thread base
+    if function.contains("UserThreadStart") || function.contains("pthread_start") {
+        return true;
+    }
+
+    false
+}
+
 fn symbolicate_stacktrace(
     thread: RawStacktrace,
     caches: &SymCacheLookup,
     signal: Option<Signal>,
-) -> CompleteStacktrace {
+) -> (CompleteStacktrace, StacktraceMetrics) {
```

Comment:
why do we have to pass this back?  I didn't spot any usage that you couldn't record here directly and you wouldn't need to complicate these signatures.  You also don't really need to declare a new struct with a bunch of implicit but undeclared metric types and update semantics as you can just call `metric!` directly which immediately shows you the metric type in the call.